src/coreclr/vm/methodtable.cpp

```diff
@@ -4467,24 +4331,91 @@ void MethodTable::CheckRunClassInitThrowing()
     // To find GC hole easier...
     TRIGGERSGC();
 
-    if (IsClassPreInited())
-        return;
-
     // Don't initialize shared generic instantiations (e.g. MyClass<__Canon>)
     if (IsSharedByGenericInstantiations())
         return;
 
-    DomainLocalModule *pLocalModule = GetDomainLocalModule();
-    _ASSERTE(pLocalModule);
+    EnsureStaticDataAllocated();
+    DoRunClassInitThrowing();
+}
+
+void MethodTable::EnsureStaticDataAllocated()
+{
+    WRAPPER_NO_CONTRACT;
+    PTR_MethodTableAuxiliaryData pAuxiliaryData = GetAuxiliaryDataForWrite();
+    if (!pAuxiliaryData->IsStaticDataAllocated() && IsDynamicStatics())
+    {
+        DynamicStaticsInfo *pDynamicStaticsInfo = GetDynamicStaticsInfo();
+        // Allocate space for normal statics if we might have them
+        if (pDynamicStaticsInfo->GetNonGCStaticsPointer() == NULL)
+            GetLoaderAllocator()->AllocateBytesForStaticVariables(pDynamicStaticsInfo, GetClass()->GetNonGCRegularStaticFieldBytes());
+
+        if (pDynamicStaticsInfo->GetGCStaticsPointer() == NULL)
+            GetLoaderAllocator()->AllocateGCHandlesBytesForStaticVariables(pDynamicStaticsInfo, GetClass()->GetNumHandleRegularStatics(), this->HasBoxedRegularStatics() ? this : NULL);
+    }
+    pAuxiliaryData->SetIsStaticDataAllocated();
+}
+
+void MethodTable::AttemptToPreinit()
+{
```

Comment:
Contract?

Reply:
My laziness strikes again. I've pulled the contract from the various allocate functions just like EnsureStaticDataAllocated. That should work